cpp/arcticdb/util/test/test_reliable_storage_lock.cpp

```diff
@@ -0,0 +1,127 @@
+    ASSERT_EQ(lock1.try_take_lock(), std::optional<uint64_t>{0});
+    Clock::time_ = 5;
+    ASSERT_EQ(lock2.try_take_lock(), std::nullopt);
+    Clock::time_ = 10;
+    ASSERT_EQ(lock2.try_take_lock(), std::nullopt);
+    Clock::time_ = 19;
+    ASSERT_EQ(lock1.try_take_lock(), std::nullopt);
+
+    // Once the first lock has expired we can take a new lock with epoch=1
+    Clock::time_ = 20;
+    ASSERT_EQ(lock2.try_take_lock(), std::optional<uint64_t>{1});
+    ASSERT_EQ(count_locks(), 2);
+
+    // We can extend the lock timeout at 25 to 35 and get an epoch=2
+    Clock::time_ = 25;
+    ASSERT_EQ(lock1.try_take_lock(), std::nullopt);
+    ASSERT_EQ(lock2.try_extend_lock(1), std::optional<uint64_t>{2});
+    ASSERT_EQ(count_locks(), 3);
+    Clock::time_ = 34;
+    ASSERT_EQ(lock1.try_take_lock(), std::nullopt);
+
+    // At time 35 the lock with epoch=2 has expired and we can re-aquire the lock
+    Clock::time_ = 35;
+    ASSERT_EQ(lock1.try_take_lock(), std::optional<uint64_t>{3});
+    ASSERT_EQ(count_locks(), 4);
+    ASSERT_EQ(lock2.try_take_lock(), std::nullopt);
+
+    // And we can free the lock immediately to allow re-aquiring without waiting for timeout
+    lock1.free_lock(3);
+    ASSERT_EQ(lock2.try_take_lock(), std::optional<uint64_t>{4});
+    // Taking lock2 with timeout=10 means we should clear all locks which have expired before 25. In this case just epoch=0
+    ASSERT_EQ(count_locks(), 4);
+
+    // But if we take a lock at 100 all locks would have expired a timeout=10 ago, and we should clear all apart from latest epoch=5
+    Clock::time_ = 100;
+    ASSERT_EQ(lock2.try_take_lock(), std::optional<uint64_t>{5});
+    ASSERT_EQ(count_locks(), 1);
+}
+
+struct SlowIncrementTask : async::BaseTask {
+    int& cnt_;
+    const ReliableStorageLock<>& lock_;
+    std::chrono::milliseconds sleep_time_;
+    bool lock_lost_ = false;
+
+    SlowIncrementTask(int& cnt, ReliableStorageLock<>& lock, std::chrono::milliseconds sleep_time) :
+        cnt_(cnt), lock_(lock), sleep_time_(sleep_time) {}
+
+    void operator()() {
+        auto guard = ReliableStorageLockGuard(lock_, [that = this](){
+            that->lock_lost_ = true;
+        });
+        auto value_before_sleep = cnt_;
+        // std::cout<<"Taken a lock with "<<value_before_sleep<<std::endl;
+        std::this_thread::sleep_for(sleep_time_);
+        if (lock_lost_) {
+            // Return early on a lost lock. We will raise an issue if this happens.
+            std::cout<<"Lost a lock with "<<value_before_sleep<<std::endl;
+            return;
+        }
+        // std::cout<<"Freeing a lock with "<<value_before_sleep<<std::endl;
+        cnt_ = value_before_sleep + 1;
+    }
+};
+
+
+TEST(ReliableStorageLock, StressMultiThreaded) {
+    // It is hard to use a piloted clock for these tests because the folly::FunctionScheduler we use for the lock
+    // extensions doesn't support a custom clock. Thus this test will need to run for about 2 minutes.
```

Comment:
It's useful to keep `test_unit_arcticdb` running fast. Might be better to add this stress test to a new cmake target. Can be done as a follow up.

Reply:
I agree it's useful to separate slower stress tests. I rememember other stress tests e.g. some of the tests in `test_slab_alloc.cpp` were also quite slow, we should probably move them together.

Adding a note to fix in the If-match cleanup.